x-pack/plugins/lens/public/datatable_visualization/expression.tsx

```diff
@@ -274,60 +314,58 @@ export const getDatatableRenderer = (dependencies: {
   },
 });
 
-function getNextOrderValue(currentValue: LensSortAction['data']['direction']) {
-  const states: Array<LensSortAction['data']['direction']> = ['asc', 'desc', 'none'];
-  const newStateIndex = (1 + states.findIndex((state) => state === currentValue)) % states.length;
-  return states[newStateIndex];
-}
+export function DatatableComponent(props: DatatableRenderProps) {
+  const [columnConfig, setColumnConfig] = useState(props.args.columns);
 
-function getDirectionLongLabel(sortDirection: LensSortAction['data']['direction']) {
-  if (sortDirection === 'none') {
-    return sortDirection;
-  }
-  return sortDirection === 'asc' ? 'ascending' : 'descending';
-}
+  useDeepCompareEffect(() => {
+    setColumnConfig(props.args.columns);
+  }, [props.args.columns]);
+  const [firstTable] = Object.values(props.data.tables);
 
-function getHeaderSortingCell(
-  name: string,
-  columnId: string,
-  sorting: Omit<LensSortAction['data'], 'action'>,
-  sortingLabel: string
-) {
-  if (columnId !== sorting.columnId || sorting.direction === 'none') {
-    return name || '';
-  }
-  // This is a workaround to hijack the title value of the header cell
-  return (
-    <span aria-sort={getDirectionLongLabel(sorting.direction)}>
-      {name || ''}
-      <EuiScreenReaderOnly>
-        <span>{sortingLabel}</span>
-      </EuiScreenReaderOnly>
-      <EuiIcon
-        className="euiTableSortIcon"
-        type={sorting.direction === 'asc' ? 'sortUp' : 'sortDown'}
-        size="m"
-        aria-label={sortingLabel}
-      />
-    </span>
+  const firstTableRef = useRef(firstTable);
+  firstTableRef.current = firstTable;
+
+  const formatFactory = props.formatFactory;
+  const formatters: Record<
+    string,
+    ReturnType<FormatFactory>
+  > = firstTableRef.current.columns.reduce(
+    (map, column) => ({
+      ...map,
+      [column.id]: formatFactory(column.meta?.params),
+    }),
+    {}
   );
-}
-
-export function DatatableComponent(props: DatatableRenderProps) {
-  const [firstTable] = Object.values(props.data.tables);
-  const formatters: Record<string, ReturnType<FormatFactory>> = {};
+  const { getType, dispatchEvent, renderMode } = props;
+  const onClickValue = useCallback(
+    (data: LensFilterEvent['data']) => {
+      dispatchEvent({ name: 'filter', data });
+    },
+    [dispatchEvent]
+  );
+  const hasAtLeastOneRowClickAction = props.rowHasRowClickTriggerActions?.find((x) => x);
```

Comment:
nitpick: there's no perf gain at all, but in general for a check I see `some` to better fit than `find` (unless the found thing is used elsewhere, but not in this case).